File: SecureStore.POS/Application/Services/ProductService.cs
using SecureStore.POS.Application.DTOs;
using SecureStore.POS.Application.Interfaces;
using SecureStore.POS.Domain.Entities;
using SecureStore.POS.Infrastructure.UnitOfWork;

namespace SecureStore.POS.Application.Services;

public class ProductService : IProductService
{
    private readonly IUnitOfWork _uow;
    private readonly IAuditService _audit;

    public ProductService(IUnitOfWork uow, IAuditService audit)
    {
        _uow = uow;
        _audit = audit;
    }

    public async Task<IEnumerable<ProductDto>> GetAllProductsAsync()
    {
        var products = await _uow.Products.GetAllActiveWithCategoryAsync();
        return products.Select(MapToDto);
    }

    public async Task<ProductDto?> GetProductByIdAsync(int id)
    {
        var p = await _uow.Products.GetByIdAsync(id);
        return p is null ? null : MapToDto(p);
    }

    public async Task<ProductDto?> GetProductByBarcodeAsync(string barcode)
    {
        var p = await _uow.Products.GetByBarcodeAsync(barcode);
        return p is null ? null : MapToDto(p);
    }

    public async Task<IEnumerable<ProductSearchResultDto>> SearchProductsAsync(string term, int maxResults = 20)
    {
        var results = await _uow.Products.SearchAsync(term, maxResults);
        return results.Select(p => new ProductSearchResultDto(
            p.ProductID, p.ProductName, p.Barcode, p.Brand,
            p.Category?.CategoryName ?? "", p.SellingPrice, p.StockQuantity));
    }

    public async Task<(bool success, string message, int productId)> CreateProductAsync(
        CreateProductDto dto, int createdByUserId)
    {
        try
        {
            var productName = dto.ProductName?.Trim() ?? string.Empty;
            var barcode = string.IsNullOrWhiteSpace(dto.Barcode) ? null : dto.Barcode.Trim();
            var brand = string.IsNullOrWhiteSpace(dto.Brand) ? null : dto.Brand.Trim();

            if (string.IsNullOrWhiteSpace(productName))
                return (false, "Product name is required.", 0);
            if (productName.Length > 200)
                return (false, "Product name cannot exceed 200 characters.", 0);
            if (barcode?.Length > 100)
                return (false, "Barcode cannot exceed 100 characters.", 0);
            if (brand?.Length > 100)
                return (false, "Brand cannot exceed 100 characters.", 0);
            if (dto.SellingPrice <= 0)
                return (false, "Selling price must be greater than 0.", 0);
            if (dto.StockQuantity < 0)
                return (false, "Stock quantity cannot be negative.", 0);
            if (dto.PurchasePrice < 0)
                return (false, "Purchase price cannot be negative.", 0);
            if (dto.ReorderLevel < 0)
                return (false, "Reorder level cannot be negative.", 0);

            var category = await _uow.Categories.GetByIdAsync(dto.CategoryID);
            if (category is null || !category.IsActive)
                return (false, "Select a valid active category.", 0);

            // Barcode uniqueness check
            if (!string.IsNullOrWhiteSpace(barcode))
            {
                var existing = await _uow.Products.ExistsAsync(
                    p => p.Barcode == barcode && p.IsActive);
                if (existing)
                    return (false, $"Barcode '{barcode}' is already assigned to another product.", 0);
            }

            var product = new Product
            {
                ProductName = productName,
                Barcode = barcode,
                CategoryID = dto.CategoryID,
                Brand = brand,
                PurchasePrice = dto.PurchasePrice,
                SellingPrice = dto.SellingPrice,
                StockQuantity = dto.StockQuantity,
                ReorderLevel = dto.ReorderLevel,
                IsActive = true,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            await _uow.Products.AddAsync(product);
            await _uow.SaveChangesAsync();

            await _audit.LogAsync("PRODUCT_CREATED", "Product", product.ProductID.ToString(),
                $"Product '{product.ProductName}' created with stock={product.StockQuantity}.",
                createdByUserId);

            return (true, "Product created successfully.", product.ProductID);
        }
        catch (Exception ex)
        {
            AppLogger.LogError(ex, "ProductService.CreateProductAsync");
            return (false, "Failed to create product. Please try again.", 0);
        }
    }

    public async Task<(bool success, string message)> UpdateProductAsync(
        UpdateProductDto dto, int updatedByUserId)
    {
        try
        {
            var product = await _uow.Products.GetByIdAsync(dto.ProductID);
            if (product is null) return (false, "Product not found.");

            var productName = dto.ProductName?.Trim() ?? string.Empty;
            var barcode = string.IsNullOrWhiteSpace(dto.Barcode) ? null : dto.Barcode.Trim();
            var brand = string.IsNullOrWhiteSpace(dto.Brand) ? null : dto.Brand.Trim();

            if (string.IsNullOrWhiteSpace(productName))
                return (false, "Product name is required.");
            if (productName.Length > 200)
                return (false, "Product name cannot exceed 200 characters.");
            if (barcode?.Length > 100)
                return (false, "Barcode cannot exceed 100 characters.");
            if (brand?.Length > 100)
                return (false, "Brand cannot exceed 100 characters.");
            if (dto.SellingPrice <= 0)
                return (false, "Selling price must be greater than 0.");
            if (dto.StockQuantity < 0)
                return (false, "Stock quantity cannot be negative.");
            if (dto.PurchasePrice < 0)
                return (false, "Purchase price cannot be negative.");
            if (dto.ReorderLevel < 0)
                return (false, "Reorder level cannot be negative.");

            var category = await _uow.Categories.GetByIdAsync(dto.CategoryID);
            if (category is null || !category.IsActive)
                return (false, "Select a valid active category.");

            // Barcode uniqueness (exclude self)
            if (!string.IsNullOrWhiteSpace(barcode))
            {
                var conflict = await _uow.Products.ExistsAsync(
                    p => p.Barcode == barcode && p.ProductID != dto.ProductID && p.IsActive);
                if (conflict)
                    return (false, $"Barcode '{barcode}' is already in use by another product.");
            }

            product.ProductName = productName;
            product.Barcode = barcode;
            product.CategoryID = dto.CategoryID;
            product.Brand = brand;
            product.PurchasePrice = dto.PurchasePrice;
            product.SellingPrice = dto.SellingPrice;
            product.StockQuantity = dto.StockQuantity;
            product.ReorderLevel = dto.ReorderLevel;
            product.IsActive = dto.IsActive;
            product.UpdatedAt = DateTime.UtcNow;

            _uow.Products.Update(product);
            await _uow.SaveChangesAsync();

            await _audit.LogAsync("PRODUCT_UPDATED", "Product", product.ProductID.ToString(),
                $"Product '{product.ProductName}' updated.", updatedByUserId);

            return (true, "Product updated successfully.");
        }
        catch (Exception ex)
        {
            AppLogger.LogError(ex, "ProductService.UpdateProductAsync");
            return (false, "Failed to update product.");
        }
    }

    public async Task<(bool success, string message)> DeactivateProductAsync(
        int productId, int deletedByUserId)
    {
        var product = await _uow.Products.GetByIdAsync(productId);
        if (product is null) return (false, "Product not found.");

        product.IsActive = false;
        product.UpdatedAt = DateTime.UtcNow;
        _uow.Products.Update(product);
        await _uow.SaveChangesAsync();

        await _audit.LogAsync("PRODUCT_DEACTIVATED", "Product", productId.ToString(),
            $"Product '{product.ProductName}' deactivated.", deletedByUserId);

        return (true, "Product deactivated successfully.");
    }

    public async Task<(bool success, string message)> ActivateProductAsync(
        int productId, int activatedByUserId)
    {
        var product = await _uow.Products.GetByIdAsync(productId);
        if (product is null) return (false, "Product not found.");
        if (product.IsActive) return (true, "Product is already active.");
        if (product.SellingPrice <= 0) return (false, "Set a valid selling price before activating this product.");
        if (product.StockQuantity < 0) return (false, "Product stock is invalid.");

        var productName = product.ProductName.Trim().ToLower();
        var brand = (product.Brand ?? string.Empty).Trim().ToLower();
        var nameConflict = await _uow.Products.ExistsAsync(p =>
            p.ProductID != product.ProductID &&
            p.IsActive &&
            p.ProductName.ToLower() == productName &&
            ((p.Brand ?? string.Empty).ToLower()) == brand);
        if (nameConflict)
            return (false, "Another active product already uses this name and brand.");

        if (!string.IsNullOrWhiteSpace(product.Barcode))
        {
            var barcodeConflict = await _uow.Products.ExistsAsync(p =>
                p.ProductID != product.ProductID &&
                p.IsActive &&
                p.Barcode == product.Barcode);
            if (barcodeConflict)
                return (false, "Barcode already exists on another active product.");
        }

        product.IsActive = true;
        product.UpdatedAt = DateTime.UtcNow;
        _uow.Products.Update(product);
        await _uow.SaveChangesAsync();

        await _audit.LogAsync("PRODUCT_ACTIVATED", "Product", productId.ToString(),
            $"Product '{product.ProductName}' activated.", activatedByUserId);

        return (true, "Product activated successfully.");
    }

    public async Task<IEnumerable<ProductDto>> GetLowStockProductsAsync()
    {
        var products = await _uow.Products.GetLowStockAsync();
        return products.Select(MapToDto);
    }

    public async Task<IEnumerable<Domain.Entities.Category>> GetCategoriesAsync() =>
        await _uow.Categories.GetAllAsync();

    private static ProductDto MapToDto(Product p) => new(
        p.ProductID, p.ProductName, p.Barcode, p.CategoryID,
        p.Category?.CategoryName ?? "", p.Brand,
        p.PurchasePrice, p.SellingPrice, p.StockQuantity,
        p.ReorderLevel, p.IsActive, p.IsLowStock,
        p.CreatedAt, p.UpdatedAt);
}

File: pos-app/electron/services/productService.ts
import { all, get, withTx, type Db } from '../database/database'
import { auditService } from './auditService'
import { backupService } from './backupService'
import { logger } from '../utils/logger'
import { PublicError, publicErrorMessage } from '../utils/safeErrors'
import type { CreateProductRequest, UpdateProductRequest, ServiceResult } from '../../shared/types'

function cleanText(value: string | undefined) {
  return String(value ?? '').trim()
}

function cleanOptionalText(value: string | undefined) {
  const text = cleanText(value)
  return text.length > 0 ? text : null
}

function toMoney(value: number) {
  return Math.round(Number(value) * 100) / 100
}

function toWholeNumber(value: number) {
  return Number(value)
}

function validateProduct(req: CreateProductRequest) {
  const productName = cleanText(req.productName)
  const brand = cleanOptionalText(req.brand)
  const barcode = cleanOptionalText(req.barcode)
  const categoryId = toWholeNumber(req.categoryId)
  const purchasePrice = toMoney(req.purchasePrice)
  const sellingPrice = toMoney(req.sellingPrice)
  const stockQuantity = toWholeNumber(req.stockQuantity)
  const reorderLevel = toWholeNumber(req.reorderLevel)

  if (!productName) return { error: 'Product name is required.' }
  if (productName.length > 200) return { error: 'Product name must be 200 characters or fewer.' }
  if (brand && brand.length > 100) return { error: 'Brand must be 100 characters or fewer.' }
  if (barcode && barcode.length > 100) return { error: 'Barcode must be 100 characters or fewer.' }
  if (!Number.isInteger(categoryId) || categoryId <= 0) return { error: 'Choose a valid category.' }
  if (!Number.isFinite(purchasePrice) || purchasePrice < 0 || purchasePrice > 1e9) return { error: 'Purchase price cannot be negative.' }
  if (!Number.isFinite(sellingPrice) || sellingPrice <= 0 || sellingPrice > 1e9) return { error: 'Selling price must be greater than 0.' }
  if (purchasePrice > sellingPrice) return { error: 'Selling price should be equal to or higher than purchase price.' }
  if (!Number.isSafeInteger(stockQuantity) || stockQuantity < 0 || stockQuantity > 100000000) return { error: 'Stock quantity cannot be negative.' }
  if (!Number.isSafeInteger(reorderLevel) || reorderLevel < 0 || reorderLevel > 100000000) return { error: 'Reorder level cannot be negative.' }

  return {
    product: {
      productName,
      brand,
      barcode,
      categoryId,
      purchasePrice,
      sellingPrice,
      stockQuantity,
      reorderLevel
    }
  }
}

async function createRequiredProductBackup(product: Record<string, unknown>, action: string, userId: number, metadata?: Record<string, unknown>, client?: Db) {
  const result = await backupService.insertProductBackup(product, action, userId, metadata, client)
  if (!result.success) throw new PublicError(result.message || 'Product backup failed.')
}

export const productService = {
  getAll: async () => {
    try {
      const data = await all(`
         SELECT p.ProductID AS "productId",
           p.ProductName AS "productName",
           p.Barcode AS "barcode",
           p.CategoryID AS "categoryId",
           c.CategoryName AS "categoryName",
           p.Brand AS "brand",
           p.PurchasePrice AS "purchasePrice",
           p.SellingPrice AS "sellingPrice",
           p.StockQuantity AS "stockQuantity",
           round(p.PurchasePrice * p.StockQuantity, 2) AS "stockCost",
           round(p.SellingPrice * p.StockQuantity, 2) AS "stockRetailValue",
           round((p.SellingPrice - p.PurchasePrice) * p.StockQuantity, 2) AS "stockProfitPotential",
           p.ReorderLevel AS "reorderLevel",
           p.IsActive AS "isActive",
           p.CreatedAt AS "createdAt",
           p.UpdatedAt AS "updatedAt",
           (p.StockQuantity <= p.ReorderLevel) AS "isLowStock"
        FROM Products p
        LEFT JOIN Categories c ON p.CategoryID = c.CategoryID
        WHERE p.IsActive = true
        ORDER BY p.ProductName ASC
      `)
      return { success: true, data, message: '' }
    } catch (error: any) {
      logger.error('Error fetching products', error)
      return { success: false, message: publicErrorMessage(error), data: [] }
    }
  },

  getAllForStock: async () => {
    try {
      const data = await all(`
         SELECT p.ProductID AS "productId",
           p.ProductName AS "productName",
           p.Barcode AS "barcode",
           p.CategoryID AS "categoryId",
           c.CategoryName AS "categoryName",
           p.Brand AS "brand",
           p.PurchasePrice AS "purchasePrice",
           p.SellingPrice AS "sellingPrice",
           p.StockQuantity AS "stockQuantity",
           round(p.PurchasePrice * p.StockQuantity, 2) AS "stockCost",
           round(p.SellingPrice * p.StockQuantity, 2) AS "stockRetailValue",
           round((p.SellingPrice - p.PurchasePrice) * p.StockQuantity, 2) AS "stockProfitPotential",
           p.ReorderLevel AS "reorderLevel",
           p.IsActive AS "isActive",
           p.CreatedAt AS "createdAt",
           p.UpdatedAt AS "updatedAt",
           (p.StockQuantity <= p.ReorderLevel) AS "isLowStock"
        FROM Products p
        LEFT JOIN Categories c ON p.CategoryID = c.CategoryID
        ORDER BY p.IsActive DESC, p.ProductName ASC
      `)
      return { success: true, data, message: '' }
    } catch (error: any) {
      logger.error('Error fetching stock products', error)
      return { success: false, message: publicErrorMessage(error), data: [] }
    }
  },

  search: async (term: string, limit = 20) => {
    try {
      const safeTerm = String(term ?? '').trim().slice(0, 100)
      const safeLimit = Number.isInteger(limit) ? Math.min(Math.max(limit, 1), 100) : 20
      const searchTerm = `%${safeTerm}%`
      const data = await all(`
         SELECT p.ProductID AS "productId", p.ProductName AS "productName",
           p.Barcode AS "barcode",
           p.Brand AS "brand",
           c.CategoryName AS "categoryName",
           p.SellingPrice AS "sellingPrice",
           p.StockQuantity AS "stockQuantity"
        FROM Products p
        LEFT JOIN Categories c ON p.CategoryID = c.CategoryID
        WHERE p.IsActive = true
          AND (p.ProductName ILIKE $1 OR p.Brand ILIKE $1 OR c.CategoryName ILIKE $1 OR p.Barcode ILIKE $1)
        ORDER BY p.ProductName ASC
        LIMIT $2
      `, [searchTerm, safeLimit])
      return { success: true, data, message: '' }
    } catch (error: any) {
      logger.error('Error searching products', error)
      return { success: false, message: publicErrorMessage(error), data: [] }
    }
  },

  getById: async (productId: number) => {
    try {
      const product = await get(`
         SELECT p.ProductID AS "productId",
           p.ProductName AS "productName",
           p.Barcode AS "barcode",
           p.CategoryID AS "categoryId",
           c.CategoryName AS "categoryName",
           p.Brand AS "brand",
           p.PurchasePrice AS "purchasePrice",
           p.SellingPrice AS "sellingPrice",
           p.StockQuantity AS "stockQuantity",
           round(p.PurchasePrice * p.StockQuantity, 2) AS "stockCost",
           round(p.SellingPrice * p.StockQuantity, 2) AS "stockRetailValue",
           round((p.SellingPrice - p.PurchasePrice) * p.StockQuantity, 2) AS "stockProfitPotential",
           p.ReorderLevel AS "reorderLevel",
           p.IsActive AS "isActive",
           p.CreatedAt AS "createdAt",
           p.UpdatedAt AS "updatedAt"
        FROM Products p
        LEFT JOIN Categories c ON p.CategoryID = c.CategoryID
        WHERE p.ProductID = $1
      `, [productId])
      return { success: true, data: product, message: '' }
    } catch (error: any) {
      return { success: false, message: publicErrorMessage(error), data: null }
    }
  },

  create: async (req: CreateProductRequest, userId: number): Promise<ServiceResult> => {
    try {
      const validated = validateProduct(req)
      if (validated.error || !validated.product) return { success: false, message: validated.error || 'Invalid product.' }
      const product = validated.product

      await withTx(async (tx: Db) => {
        const category = await tx.get('SELECT 1 FROM Categories WHERE CategoryID = $1 AND IsActive = true', [product.categoryId])
        if (!category) throw new PublicError('Choose an active category.')

        const existing = await tx.get(`
          SELECT ProductID FROM Products
          WHERE IsActive = true
            AND lower(ProductName) = lower($1)
            AND coalesce(lower(Brand), '') = coalesce(lower($2), '')
        `, [product.productName, product.brand])
        if (existing) throw new PublicError('An active product with this name and brand already exists.')

        if (product.barcode) {
          const barcodeExists = await tx.get('SELECT ProductID FROM Products WHERE Barcode = $1', [product.barcode])
          if (barcodeExists) throw new PublicError('A product with this barcode already exists.')
        }

        const info = await tx.run(
          `INSERT INTO Products (ProductName, CategoryID, Brand, Barcode, PurchasePrice, SellingPrice, StockQuantity, ReorderLevel)
           VALUES ($1, $2, $3, $4, $5, $6, $7, $8) RETURNING ProductID`,
          [product.productName, product.categoryId, product.brand, product.barcode, product.purchasePrice, product.sellingPrice, product.stockQuantity, product.reorderLevel]
        )
        const productId = info.rows[0].ProductID

        if (product.stockQuantity > 0) {
          await tx.run(
            `INSERT INTO InventoryTransactions (ProductID, TransactionType, QuantityChange, OldStock, NewStock, UserID, Reason)
             VALUES ($1, 'Purchase', $2, 0, $2, $3, 'Opening stock entered when product was created')`,
            [productId, product.stockQuantity, userId]
          )
        }
        await auditService.log('PRODUCT_CREATED', 'Product', `Created product ${product.productName}`, userId, String(productId), tx)
        const created = await tx.get('SELECT ProductID as productId, ProductName, Barcode, CategoryID as categoryId, Brand, PurchasePrice as purchasePrice, SellingPrice as sellingPrice, StockQuantity as stockQuantity, ReorderLevel as reorderLevel, IsActive, CreatedAt, UpdatedAt FROM Products WHERE ProductID = $1', [productId])
        await createRequiredProductBackup(created as any, 'CREATE', userId, { after: created }, tx)
      })

      return { success: true, message: 'Product created successfully' }
    } catch (error: any) {
      logger.error('Error creating product', error)
      return { success: false, message: publicErrorMessage(error) }
    }
  },

  update: async (req: UpdateProductRequest, userId: number): Promise<ServiceResult> => {
    try {
      const productId = toWholeNumber(req.productId)
      if (!Number.isInteger(productId) || productId <= 0) return { success: false, message: 'Invalid product.' }

      const validated = validateProduct(req)
      if (validated.error || !validated.product) return { success: false, message: validated.error || 'Invalid product.' }
      const product = validated.product

      await withTx(async (tx: Db) => {
        const current = await tx.get('SELECT ProductName, Barcode, CategoryID, Brand, PurchasePrice, SellingPrice, StockQuantity, ReorderLevel, IsActive FROM Products WHERE ProductID = $1', [productId]) as any
        if (!current) throw new PublicError('Product was not found.')

        const category = await tx.get('SELECT 1 FROM Categories WHERE CategoryID = $1 AND IsActive = true', [product.categoryId])
        if (!category) throw new PublicError('Choose an active category.')

        const existing = await tx.get(`
          SELECT ProductID FROM Products
          WHERE ProductID != $1
            AND IsActive = true
            AND lower(ProductName) = lower($2)
            AND coalesce(lower(Brand), '') = coalesce(lower($3), '')
        `, [productId, product.productName, product.brand])
        if (existing) throw new PublicError('Another active product already uses this name and brand.')

        if (product.barcode) {
          const barcodeExists = await tx.get('SELECT ProductID FROM Products WHERE ProductID != $1 AND Barcode = $2', [productId, product.barcode])
          if (barcodeExists) throw new PublicError('Another product already uses this barcode.')
        }

        await tx.run(
          `UPDATE Products
           SET ProductName = $1, CategoryID = $2, Brand = $3, Barcode = $4,
               PurchasePrice = $5, SellingPrice = $6, StockQuantity = $7,
               ReorderLevel = $8, IsActive = $9, UpdatedAt = now()
           WHERE ProductID = $10`,
          [
            product.productName,
            product.categoryId,
            product.brand,
            product.barcode,
            product.purchasePrice,
            product.sellingPrice,
            product.stockQuantity,
            product.reorderLevel,
            Boolean(req.isActive),
            productId
          ]
        )

        if (Number(current.StockQuantity) !== product.stockQuantity) {
          await tx.run(
            `INSERT INTO InventoryTransactions (ProductID, TransactionType, QuantityChange, OldStock, NewStock, UserID, Reason)
             VALUES ($1, 'Adjustment', $2, $3, $4, $5, 'Stock corrected from product edit')`,
            [
              productId,
              product.stockQuantity - Number(current.StockQuantity),
              Number(current.StockQuantity),
              product.stockQuantity,
              userId
            ]
          )
        }

        const changes: string[] = []
        if (current.ProductName !== product.productName) changes.push(`productName: ${current.ProductName} -> ${product.productName}`)
        if ((current.Barcode || null) !== (product.barcode || null)) changes.push(`barcode: ${current.Barcode || ''} -> ${product.barcode || ''}`)
        if (Number(current.CategoryID) !== product.categoryId) changes.push(`categoryId: ${current.CategoryID} -> ${product.categoryId}`)
        if ((current.Brand || null) !== (product.brand || null)) changes.push(`brand: ${current.Brand || ''} -> ${product.brand || ''}`)
        if (Number(current.PurchasePrice) !== product.purchasePrice) changes.push(`purchasePrice: ${current.PurchasePrice} -> ${product.purchasePrice}`)
        if (Number(current.SellingPrice) !== product.sellingPrice) changes.push(`sellingPrice: ${current.SellingPrice} -> ${product.sellingPrice}`)
        if (Number(current.StockQuantity) !== product.stockQuantity) changes.push(`stockQuantity: ${current.StockQuantity} -> ${product.stockQuantity}`)
        if (Number(current.ReorderLevel) !== product.reorderLevel) changes.push(`reorderLevel: ${current.ReorderLevel} -> ${product.reorderLevel}`)
        if (Boolean(current.IsActive) !== Boolean(req.isActive)) changes.push(`isActive: ${current.IsActive} -> ${req.isActive}`)
        await auditService.log('PRODUCT_UPDATED', 'Product', `Updated product ${product.productName}. ${changes.join('; ')}`, userId, productId.toString(), tx)
        const updated = await tx.get('SELECT ProductID as productId, ProductName, Barcode, CategoryID as categoryId, Brand, PurchasePrice as purchasePrice, SellingPrice as sellingPrice, StockQuantity as stockQuantity, ReorderLevel as reorderLevel, IsActive, CreatedAt, UpdatedAt FROM Products WHERE ProductID = $1', [productId])
        await createRequiredProductBackup(updated as any, 'UPDATE', userId, { before: current, after: updated }, tx)
      })

      return { success: true, message: 'Product updated successfully' }
    } catch (error: any) {
      logger.error('Error updating product', error)
      return { success: false, message: publicErrorMessage(error) }
    }
  },

  deactivate: async (productId: number, userId: number): Promise<ServiceResult> => {
    try {
      const safeProductId = toWholeNumber(productId)
      if (!Number.isInteger(safeProductId) || safeProductId <= 0) return { success: false, message: 'Invalid product.' }

      const message = await withTx(async (tx: Db) => {
        const current = await tx.get('SELECT ProductID as productId, ProductName, Barcode, CategoryID as categoryId, Brand, PurchasePrice as purchasePrice, SellingPrice as sellingPrice, StockQuantity as stockQuantity, ReorderLevel as reorderLevel, IsActive, CreatedAt, UpdatedAt FROM Products WHERE ProductID = $1', [safeProductId]) as any
        if (!current) throw new PublicError('Product was not found.')
        if (!current.IsActive) return 'Product is already inactive.'

        await tx.run('UPDATE Products SET IsActive = false, UpdatedAt = now() WHERE ProductID = $1', [safeProductId])
        await auditService.log('PRODUCT_DEACTIVATED', 'Product', `Deactivated product ${current.ProductName} (ProductID ${safeProductId})`, userId, safeProductId.toString(), tx)
        const deactivated = await tx.get('SELECT ProductID as productId, ProductName, Barcode, CategoryID as categoryId, Brand, PurchasePrice as purchasePrice, SellingPrice as sellingPrice, StockQuantity as stockQuantity, ReorderLevel as reorderLevel, IsActive, CreatedAt, UpdatedAt FROM Products WHERE ProductID = $1', [safeProductId])
        await createRequiredProductBackup(deactivated as any, 'DEACTIVATE', userId, { before: current, after: deactivated }, tx)
        return 'Product deactivated and removed from checkout successfully'
      })

      return { success: true, message }
    } catch (error: any) {
      return { success: false, message: publicErrorMessage(error) }
    }
  },

  activate: async (productId: number, userId: number): Promise<ServiceResult> => {
    try {
      const safeProductId = toWholeNumber(productId)
      if (!Number.isInteger(safeProductId) || safeProductId <= 0) return { success: false, message: 'Invalid product.' }

      const message = await withTx(async (tx: Db) => {
        const current = await tx.get('SELECT ProductID as productId, ProductName, Barcode, CategoryID as categoryId, Brand, PurchasePrice as purchasePrice, SellingPrice as sellingPrice, StockQuantity as stockQuantity, ReorderLevel as reorderLevel, IsActive, CreatedAt, UpdatedAt FROM Products WHERE ProductID = $1', [safeProductId]) as any
        if (!current) throw new PublicError('Product was not found.')
        if (current.IsActive) return 'Product is already active in checkout.'
        if (Number(current.SellingPrice) <= 0) throw new PublicError('Set a valid selling price before activating this product.')
        if (Number(current.StockQuantity) < 0) throw new PublicError('Product stock is invalid.')

        const duplicate = await tx.get(`
          SELECT ProductID FROM Products
          WHERE ProductID != $1
            AND IsActive = true
            AND lower(ProductName) = lower($2)
            AND coalesce(lower(Brand), '') = coalesce(lower($3), '')
        `, [safeProductId, current.ProductName, current.Brand])
        if (duplicate) throw new PublicError('Another active product already uses this name and brand.')

        await tx.run('UPDATE Products SET IsActive = true, UpdatedAt = now() WHERE ProductID = $1', [safeProductId])
        await auditService.log('PRODUCT_ACTIVATED', 'Product', `Activated product ${current.ProductName} (ProductID ${safeProductId})`, userId, safeProductId.toString(), tx)
        const activated = await tx.get('SELECT ProductID as productId, ProductName, Barcode, CategoryID as categoryId, Brand, PurchasePrice as purchasePrice, SellingPrice as sellingPrice, StockQuantity as stockQuantity, ReorderLevel as reorderLevel, IsActive, CreatedAt, UpdatedAt FROM Products WHERE ProductID = $1', [safeProductId])
        await createRequiredProductBackup(activated as any, 'ACTIVATE', userId, { before: current, after: activated }, tx)
        return 'Product activated and returned to checkout successfully.'
      })

      return { success: true, message }
    } catch (error: any) {
      logger.error('Error activating product', error)
      return { success: false, message: publicErrorMessage(error) }
    }
  },

  getLowStock: async () => {
    try {
      const data = await all(`
        SELECT p.ProductID as productId,
               p.ProductName as productName,
               p.Barcode as barcode,
               p.CategoryID as categoryId,
               c.CategoryName as categoryName,
               p.Brand as brand,
               p.PurchasePrice as purchasePrice,
               p.SellingPrice as sellingPrice,
               p.StockQuantity as stockQuantity,
               round(p.PurchasePrice * p.StockQuantity, 2) as stockCost,
               round(p.SellingPrice * p.StockQuantity, 2) as stockRetailValue,
               round((p.SellingPrice - p.PurchasePrice) * p.StockQuantity, 2) as stockProfitPotential,
               p.ReorderLevel as reorderLevel,
               p.IsActive as isActive,
               p.CreatedAt as createdAt,
               p.UpdatedAt as updatedAt
        FROM Products p
        LEFT JOIN Categories c ON p.CategoryID = c.CategoryID
        WHERE p.IsActive = true AND p.StockQuantity <= p.ReorderLevel
        ORDER BY p.StockQuantity ASC
      `)
      return { success: true, data, message: '' }
    } catch (error: any) {
      return { success: false, message: publicErrorMessage(error), data: [] }
    }
  },

  getAllCategories: async () => {
    try {
      const data = await all(`
         SELECT CategoryID AS "categoryId",
           CategoryName AS "categoryName",
           Description AS "description",
           IsActive AS "isActive"
        FROM Categories
        WHERE IsActive = true
        ORDER BY CategoryName
      `)
      return { success: true, data, message: '' }
    } catch (error: any) {
      return { success: false, message: publicErrorMessage(error), data: [] }
    }
  }
}
